python/letsql/expr/api.py

```diff
@@ -1655,3 +1659,96 @@ def to_parquet(expr: ir.Expr, path: str | Path, **kwargs: Any):
             con.register(t.to_expr(), t.name)
 
     return con.to_parquet(expr, path, **kwargs)
+
+
+def train_test_split(
+    table: ir.Table,
+    unique_key: str | list[str],
+    test_size: float = 0.25,
+    num_buckets: int = 100,
+    random_seed: int | None = None,
+) -> tuple[ir.Table, ir.Table]:
+    """Randomly split Ibis table data into training and testing tables.
+
+    This function splits an Ibis table into training and testing tables
+    based on a unique key or combination of keys. It uses a hashing function to
+    convert the unique key into an integer, then applies a modulo operation to split
+    the data into buckets. The training table consists of data points from a subset of
+    these buckets, while the remaining data points form the test table.
+
+    Parameters
+    ----------
+    table
+        The input Ibis table to be split.
+    unique_key
+        The column name(s) that uniquely identify each row in the table. This unique_key
+        is used to create a deterministic split of the dataset through a hashing
+        process.
+    test_size
+        The ratio of the dataset to include in the test split, which should be between
+        0 and 1. This ratio is approximate because the hashing algorithm may not provide
+        a uniform bucket distribution for small datasets. Larger datasets will result in
+        more uniform bucket assignments, making the split ratio closer to the desired
+        value.
+    num_buckets
+        The number of buckets into which the data is divided during the splitting
+        process. It controls how finely the data is divided into buckets during
+        the split process. Adjusting num_buckets can affect the granularity and
+        efficiency of the splitting operation, balancing between accuracy and
+        computational efficiency.
+    random_seed
+        Seed for the random number generator. If provided, ensures reproducibility
+        of the split.
+
+    Returns
+    -------
+    tuple[ir.Table, ir.Table]
+        A tuple containing two Ibis tables: (train_table, test_table).
+
+    Raises
+    ------
+    ValueError
+        If test_size is not a float between 0 and 1.
+
+    Examples
+    --------
+    >>> import letsql as ls
+
+    Split an Ibis table into training and testing tables.
+
+    >>> table = ibis.memtable({"key1": range(100)})
+    >>> train_table, test_table = ls.train_test_split(
+    ...     table,
+    ...     unique_key="key1",
+    ...     test_size=0.2,
+    ...     random_seed=0,
+    ... )
+    """
+    if not (0 < test_size < 1):
+        raise ValueError("test size should be a float between 0 and 1.")
+
+    # Set the random seed for reproducibility
+    if random_seed:
+        rndm_seed(random_seed)
+
+    # Generate a random 256-bit key
+    random_str = str(grb(256))
```

Comment:
I think this is better expressed as
`str(random.Random(random_seed).getrandbits(256))`
so that it cannot be impacted by anything outside

Reply:
I think we should sure that the seed can still use system entropy? 


```python
  # Set the random seed, if needed  for reproducibility
  if random_seed:
      random_str = str(random.Random(random_seed).getrandbits(256))
  else:
      #use system entropy to get seed os.urandom() 
      random_str = str(random.Random().getrandbits(256))

```
Users may want to rely on randomness generated  and not rely on providing a seed because the seed provided may not provide enough entropy to be random. 

I think this  option should be there , vs something where we pick a default seed. 


My preference is that we

Reply:
also it still has a scoping  issue with 
https://github.com/letsql/letsql/blob/497e33364dcf81b645962cc1c52c20f1c1b78dbb/python/letsql/expr/api.py#L460

Reply:
we can do this though `__import__("random").Random().getrandbits(1)`

Reply:
or use `importlib`.

Reply:
when `random_seed` is `None`, `random.Random()` and `random.Random(random_seed)` are the same
how about `from random import Random`?

Reply:
on second thought, i think we should avoid putting implementation into `api.py`. how about we put this somewhere else and expose it through `api.py`?

Reply:
>when random_seed is None, random.Random() and random.Random(random_seed)
Ahh yeah true .

I can move it somehwere else.